Add tests for H5R get name APIs

bmribler opened the pull request:
This PR adds functionality tests for the following APIs:
    H5Rget_file_name
    H5Rget_obj_name
    H5Rget_attr_name

It also removes a "+1" when returning a name length in H5R__get_attr_name() in the Fortran wrapper.  This extra "+1" gave an incorrect value for the length of the referenced object's attribute name.  In addition, the PR provides some minor improvement in readability and maintainability.

(Fixed GH-4447)

bmribler committed:
Add tests for H5R get name APIs

Added functionality tests for the following APIs:
H5Rget_file_name
H5Rget_obj_name
H5Rget_attr_name

Also removed "+1" when returning a name length in H5R__get_attr_name().
The exter "+1" gave an incorrect value for the length of the referenced
object's attribute name.

Fixed GH-4447

github-actions[bot] committed:
Committing clang-format changes

bmribler committed:
Fix typo

bmribler committed:
Merge branch 'fix_GH-4447' of https://github.com/bmribler/hdf5_bmr23 into fix_GH-4447

bmribler committed:
Fix typo

bmribler committed:
Fix Fortran test

The C API H5Rget_attr_name incorrectly added 1 to the length of the
referenced object's attribute name, so the Fortran API h5rget_attr_name_f
removed 1 from the returned value to accommodate the incorrectness.
This PR fixes H5Rget_attr_name so this workaround in h5rget_attr_name_f
is no longer needed.

jhendersonHDF commented:
Note that `H5R__get_file_name` also appears to add a +1. We should add a test that compares the length returned by `H5Fget_name` vs. `H5Rget_file_name` to make sure they're consistent, as well as adding a test for `H5Aget_name` vs. `H5Rget_attr_name` and `H5Iget_name` vs. `H5Rget_obj_name`.

brtnfld commented on fortran/src/H5Rff.F90:
```diff
@@ -1180,8 +1180,6 @@ END FUNCTION H5Rget_attr_name
        name_len = H5Rget_attr_name(ref_ptr, c_name, 1_SIZE_T)
        IF(name_len.LT.0_SIZE_T) hdferr = H5I_INVALID_HID_F
-       ! Don't include the NULL term in the size
-       name_len = name_len - 1
```
To clarify, H5Rget_attr_name returned length does not include the NULL character now with this PR? If that is the case, the doxygen doc needs to be updated as it says it does include NULL.

bmribler replied:
Done

brtnfld commented:
PR also needs a RELEASE.txt entry detailing the API behavior change introduced.

brtnfld commented:
> Note that `H5R__get_file_name` also appears to add a +1. We should add a test that compares the length returned by `H5Fget_name` vs. `H5Rget_file_name` to make sure they're consistent, as well as adding a test for `H5Aget_name` vs. `H5Rget_attr_name` and `H5Iget_name` vs. `H5Rget_obj_name`.

Hmm, the documentation says that the length returned does not include the null terminator for H5Rget_file_name.

jhendersonHDF commented:
> > Note that `H5R__get_file_name` also appears to add a +1. We should add a test that compares the length returned by `H5Fget_name` vs. `H5Rget_file_name` to make sure they're consistent, as well as adding a test for `H5Aget_name` vs. `H5Rget_attr_name` and `H5Iget_name` vs. `H5Rget_obj_name`.
> 
> Hmm, the documentation says that the length returned does not include the null terminator for H5Rget_file_name.

I believe most of the "name" functions currently return the length without the null terminator. But this does add a little mental effort since developers have to remember to add that in. I also believe there used to be some other discrepancies between the documentation and what the function actually returned, but those may have been resolved by now. Always good to check that all the "name" functions are correct and consistent though.

brtnfld commented:
Is the documentation wrong for H5Rget_file_name in that len does include the null termination? or is +1 in the double "_" API not related?

jhendersonHDF commented:
> Is the documentation wrong for H5Rget_file_name in that len does include the null termination? or is +1 in the double "_" API not related?

I believe the documentation is currently wrong in that it appears the length returned by the native VOL for `H5Rget_file_name` would include the NUL terminator (I haven't tested this). However, since most of the "name" functions don't include the terminator, it seems we should remove the +1 so the documentation is correct and the function would be consistent with other "name" functions.

brtnfld commented:
I agree; consistency would be good. I'm just wondering if such API changes should be made in a minor release.

bmribler commented:
> > Note that `H5R__get_file_name` also appears to add a +1. We should add a test that compares the length returned by `H5Fget_name` vs. `H5Rget_file_name` to make sure they're consistent, as well as adding a test for `H5Aget_name` vs. `H5Rget_attr_name` and `H5Iget_name` vs. `H5Rget_obj_name`.
> 
> Hmm, the documentation says that the length returned does not include the null terminator for H5Rget_file_name.

I accidentally deleted the test results so I don't remember exact detail, but I remember clearly the length returned had exact number of characters in the name, so no null terminator.  That was why the Fortran test failed in that PR before I committed the fix.

jhendersonHDF commented:
> > > Note that `H5R__get_file_name` also appears to add a +1. We should add a test that compares the length returned by `H5Fget_name` vs. `H5Rget_file_name` to make sure they're consistent, as well as adding a test for `H5Aget_name` vs. `H5Rget_attr_name` and `H5Iget_name` vs. `H5Rget_obj_name`.
> > 
> > 
> > Hmm, the documentation says that the length returned does not include the null terminator for H5Rget_file_name.
> 
> I accidentally deleted the test results so I don't remember exact detail, but I remember clearly the length returned had exact number of characters in the name, so no null terminator. That was why the Fortran test failed in that PR before I committed the fix.

It looks like the case where `H5R__get_file_name` is called is with unopened external references so the test case probably wasn't going down that path. We would need a more specific test case to hit the issue.

bmribler commented:
> > > > Note that `H5R__get_file_name` also appears to add a +1. We should add a test that compares the length returned by `H5Fget_name` vs. `H5Rget_file_name` to make sure they're consistent, as well as adding a test for `H5Aget_name` vs. `H5Rget_attr_name` and `H5Iget_name` vs. `H5Rget_obj_name`.
> > > 
> > > 
> > > Hmm, the documentation says that the length returned does not include the null terminator for H5Rget_file_name.
> > 
> > 
> > I accidentally deleted the test results so I don't remember exact detail, but I remember clearly the length returned had exact number of characters in the name, so no null terminator. That was why the Fortran test failed in that PR before I committed the fix.
> 
> It looks like the case where `H5R__get_file_name` is called is with unopened external references so the test case probably wasn't going down that path. We would need a more specific test case to hit the issue.

OK

bmribler committed:
Merge branch 'develop' of https://github.com/HDFGroup/hdf5 into fix_GH-4447

brtnfld commented:
> I accidentally deleted the test results so I don't remember exact detail, but I remember clearly the length returned had exact number of characters in the name, so no null terminator. That was why the Fortran test failed in that PR before I committed the fix.

Are you talking about H5Rget_file_name? Otherwise, originally, the length of `H5Rget_attr_name` did include the null terminator, which Fortran was subtracting from the length returned by the function.

bmribler commented:
> > I accidentally deleted the test results so I don't remember exact detail, but I remember clearly the length returned had exact number of characters in the name, so no null terminator. That was why the Fortran test failed in that PR before I committed the fix.
> 
> Are you talking about H5Rget_file_name? Otherwise, originally, the length of `H5Rget_attr_name` did include the null terminator, which Fortran was subtracting from the length returned by the function.

Yes, I found that `H5Rget_attr_name` included null terminator while other get name functions I checked didn't, so I removed +1 there to be consistent with the other getname functions.  But after I removed +1 in `H5Rget_attr_name`, the Fortran test failed, hence, I made the change there.

So, what is the decision about the null terminator for getname functions?  Should I get a tally of the current way to help us decide?  I will update the documentation where necessary.

brtnfld commented:
I think this PR addresses the immediate issue; the get_name issue can be looked into later.

bmribler commented:
> I think this PR addresses the immediate issue; the get_name issue can be looked into later.

Ok, good, thanks!  I'll make sure that the documentation will say no null terminator for these particular functions and add a RELEASE.txt entry.

BTW, the comment "I'm just wondering if such API changes should be made in a minor release." has not been discussed...  Should this not go in 1.14.6?

lrknox commented:
> I agree; consistency would be good. I'm just wondering if such API changes should be made in a minor release.

@brtnfld, were you referring to API changes in this PR, or to other potential changes?

lrknox commented:
> > I think this PR addresses the immediate issue; the get_name issue can be looked into later.
> 
> Ok, good, thanks! I'll make sure that the documentation will say no null terminator for these particular functions and add a RELEASE.txt entry.
> 
> BTW, the comment "I'm just wondering if such API changes should be made in a minor release." has not been discussed... Should this not go in 1.14.6?

It looks like Scot was ok with the PR, though he didn't officially approve it, and noone answered this question.  The PR has 3 approvals;  if noone is requests changes by tomorrow I think we should merge this.

brtnfld commented:
I think @bmribler is going to update the Doxygen documentationwill update the Doxygen documentation and release notes in the PR before it gets merged.

bmribler commented:
> I think @bmribler is going to update the Doxygen documentationwill update the Doxygen documentation and release notes in the PR before it gets merged.

Yes, I'll do it today or tomorrow.

bmribler commented:
@jhendersonHDF @brtnfld @qkoziol
Please look at issue #4704 and let me know if we still want to remove the null terminator from H5Rget_attr_name in this PR.  Or should I check the code to verify the doc info in #4704 first, before deciding about H5Rget_attr_name here?

I think removing the null terminator in this PR is a good change and the decision seemed to have been made, but now I'm a little concerned about making the inconsistency issue worse later, so I'm bringing this up again to be sure that we have a consensus.  Thanks!

brtnfld commented:
I think the consensus going forward was to never include the null terminator, so this PR should be consistent with future changes.

bmribler commented:
> I think the consensus going forward was to never include the null terminator, so this PR should be consistent with future changes.

Thanks!  I'll go ahead and wrap this up.

qkoziol commented:
> I think the consensus going forward was to never include the null terminator, so this PR should be consistent with future changes.

Agree.   These routines should behave like strlen().

bmribler committed:
Add test H5Aget_name against H5Rget_attr_name

github-actions[bot] committed:
Committing clang-format changes

bmribler committed:
Update documentation

bmribler committed:
Fix typo

brtnfld commented on release_docs/RELEASE.txt:
```diff
@@ -811,6 +811,15 @@ Bug Fixes since HDF5-1.14.0 release
+      without the null terminator.  H5Rget_attr_name now behaves consistently
+      with the other two APIs.  Going forward, all the get character string
+      APIs in HDF5 will be modified/written in this manner, regarding the
+      length of a character string.
```
remove the comma before "regarding".
